### PR TITLE
Create CheckBox component aligned with design system

### LINKED ISSUE #2702: (SP: 3) Create CheckBox component aligned with design system
#### Description:
Implement a reusable `CheckBox` component that meets the design system standards. This component will be used throughout the application to maintain consistency in checkbox styles, behavior, and accessibility.

[Figma](https://www.figma.com/design/liJZDYhUnDP15Pi9bipDcv/Space2Study-(Students)?node-id=20967-98389&t=l5PI8JRXpNFKQBr0-4)

#### Requirements:
- **Styles:**
  - Follow design system variables for colors, typography, padding, and other styles.
  - Support different variants (check, middle).
  - Ensure hover, focus, active, and disabled states are defined according to the design system.
  - Provide a label for checkbox 
  
- **Props:**
  - `variant`: Defines type (check, middle).
  - `label`: Provide a label for the checkbox.
  - `labelPosition`: Supports different positions (bottom, top, end)
  -  `color`: Supports different colors (primary, secondary, error, success)
  - `size`: Supports different sizes (sm, md, lg).
  - `loading`: Disables the checkbox when true. When loading, the  checkbox  shouldn't shrink or change its size compared to its default state
  - `disabled`: Disables the checkbox when true.

Don't forget to cover component with tests and add to the storybook

### PR DESCRIPTION
**Pull request by nebby2105**
#2702 
Created Checkbox component with the following configurable props:

1.  color (primary, secondary, success, error)
2.  variant (check, middle)
3.  label
4.  labelPosition (top, bottom, end)
5.  size (sm, md, lg)
6.  loading
7.  disabled

Default Parameters:

Color: primary
Variant: check
Size: md
Label Position: end
Loading: false
Disabled: false

Default parameters:
Screenshot 2024-11-21 at 18 17 49

Color="success"
Screenshot 2024-11-21 at 18 16 50

Color="error"
Screenshot 2024-11-21 at 18 17 04

labelPosition="top"
Screenshot 2024-11-21 at 18 17 19

labelPosition="bottom"
Screenshot 2024-11-21 at 18 17 32

- [ ] create CheckBox component 
- [ ] add it to storybook
- [ ]  cover with tests

### CODE REVIEW

File: src/design-system/components/button/Button.tsx [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+  endIcon?: ReactNode
+  to?: string
+}
+
+export type ButtonProps = BaseButtonProps &
+  Omit<MuiButtonProps, keyof BaseButtonProps>
+
+type Ref = MuiButtonProps['ref']
+
+const Button = forwardRef(
```
Comment (ShadowOfTheSpace) [inline src/design-system/components/button/Button.tsx:37]: Add generic types to `forwardRef` function. So you will not have to type `Ref` manually. Also props at 40-48 lines will be typed as well.

File: src/components/check-box/CheckBox.tsx [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  color = 'primary',
+  disabled = false,
+  label,
+  labelPosition = 'end',
+  loading = false,
+  variant = 'check',
+  size = 'md',
+  ...props
+}) => {
+  const [checked, setChecked] = useState(false)
```
Comment (ShadowOfTheSpace) [inline src/components/check-box/CheckBox.tsx]: Add generic here. 
It is a good practice to have generic type in useState hook, even if it has default value

File: src/components/check-box/CheckBox.tsx [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+          checked={checked}
+          className='s2s-checkbox__input'
+          color={color}
+          data-testid='checkbox-input'
+          disabled={disabled || loading}
+          indeterminate={variant === 'middle' && checked}
+          onChange={handleChange}
+        />
+      )}
+
```
Comment (ShadowOfTheSpace) [inline src/components/check-box/CheckBox.tsx]:
```suggestion
```

File: src/stories/CheckBox.stories.tsx [hunk 4]
```diff
@@ -0,0 +1,169 @@
+import type { Meta, StoryObj } from '@storybook/react'
+// import CheckBox from '~/components/check-box/CheckBox'
```
Comment (ShadowOfTheSpace) [inline src/stories/CheckBox.stories.tsx]:
```suggestion
```

File: src/stories/CheckBox.stories.tsx [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+}
+export default meta
+
+type Story = StoryObj<typeof CheckBox>
+
+// export const All: Story = {
+//     render: (args) => {
+
+//     }
+// }
```
Comment (ShadowOfTheSpace) [inline src/stories/CheckBox.stories.tsx]:
```suggestion
```

File: src/design-system/components/checkbox/CheckBox.tsx [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+          checked={checked}
+          className='s2s-checkbox__input'
+          color={color}
+          data-testid='checkbox-input'
+          disabled={disabled || loading}
+          indeterminate={variant === 'middle' && checked}
+          onChange={handleChange}
+        />
+      )}
+
```
Comment (ShadowOfTheSpace) [inline src/design-system/components/checkbox/CheckBox.tsx:68]:
```suggestion
```

File: src/design-system/components/checkbox/CheckBox.tsx [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+  color = 'primary',
+  disabled = false,
+  label,
+  labelPosition = 'end',
+  loading = false,
+  variant = 'check',
+  size = 'md',
+  ...props
+}) => {
+  const [checked, setChecked] = useState(false)
```
Comment (ShadowOfTheSpace) [inline src/design-system/components/checkbox/CheckBox.tsx:26]: Add generic here.
It is a good practice to have generic type in useState hook, even if it has default value

**Comment by nebby2105** [comment]
add new pr

**Pull request closed without merging**